website/identifiers/clients/crossref.py

```diff
@@ -0,0 +1,211 @@
+# -*- coding: utf-8 -*-
+import furl
+import lxml
+import time
+import logging
+
+import requests
+
+from website.identifiers.metadata import remove_control_characters
+from website.identifiers.clients.base import AbstractIdentifierClient
+from website import settings
+
+logger = logging.getLogger(__name__)
+
+CROSSREF_NAMESPACE = 'http://www.crossref.org/schema/4.4.1'
+CROSSREF_SCHEMA_LOCATION = 'http://www.crossref.org/schema/4.4.1 http://www.crossref.org/schemas/crossref4.4.1.xsd'
+CROSSREF_ACCESS_INDICATORS = 'http://www.crossref.org/AccessIndicators.xsd'
+CROSSREF_RELATIONS = 'http://www.crossref.org/relations.xsd'
+CROSSREF_SCHEMA_VERSION = '4.4.1'
+JATS_NAMESPACE = 'http://www.ncbi.nlm.nih.gov/JATS1'
+XSI = 'http://www.w3.org/2001/XMLSchema-instance'
+CROSSREF_DEPOSITOR_NAME = 'Open Science Framework'
+
+
+class CrossRefClient(AbstractIdentifierClient):
+
+    def __init__(self, base_url):
+        self.base_url = base_url
+
+    def build_doi(self, preprint):
+        from osf.models import PreprintProvider
+
+        prefix = preprint.provider.doi_prefix or PreprintProvider.objects.get(_id='osf').doi_prefix
+        return settings.DOI_FORMAT.format(prefix=prefix, guid=preprint._id)
+
+    def build_metadata(self, preprint, status='public', **kwargs):
+        """Return the crossref metadata XML document for a given preprint as a string for DOI minting purposes
+
+        :param preprint -- the preprint, or list of preprints to build metadata for
+        """
+        is_batch = False
+        if isinstance(preprint, list):
+            is_batch = True
+            preprints = preprint
+        else:
+            preprints = [preprint]
+
+        element = lxml.builder.ElementMaker(nsmap={
+            None: CROSSREF_NAMESPACE,
+            'xsi': XSI},
+        )
+
+        timestamp = str(int(time.time()))
+        batch_id = timestamp if is_batch else preprint._id
```

Comment:
This is unused now, right? Remove?

Reply:
Ah, ok, thanks for clarifying. The timestamp still shouldn't be necessary, right? Would it make sense to rename `batch_id` -> `guids` and return a comma-separated list of GUIDs like we discussed before? IIRC, there was a size limit for the tag, but we shouldn't run into that if your script only sends 5 preprints at a time. To be safe, you could just put the first 10 or so GUIDs in the tag.

Reply:
Right, we have to name it batch_id because of the fields crossref is parsing for, and indeed there was a size limit. We could stick to sending 5 preprints at a time, but since the batch limit on their end is 10,000 I figured we'd want to send a lot more than that just so the process gets done quicker! 

That's a good point though, I should change that setting in this PR to something more like 100 or 1000 (or make it configurable when you run the script? Which is better?)

Reply:
Edit: Ok I think I see what you mean -- only send the first 5 GUIDs in the tag regardless of the number of records being sent?

Reply:
Mhm, we can/should probably send more than 5 at a time, but yeah, you can limit the number of GUIDS listed in `<batch_id>` to prevent going over the size limit.